src/Compiler/Checking/PostInferenceChecks.fs

```diff
@@ -302,12 +302,24 @@ let BindVal cenv env (v: Val) =
     //printfn "binding %s..." v.DisplayName
     let alreadyDone = cenv.boundVals.ContainsKey v.Stamp
     cenv.boundVals[v.Stamp] <- 1
+    
+    let topLevelPublicLetBindingNotInSignatureFile () =
+        let parentHasSignatureFile () =
+            match v.TryDeclaringEntity with
+            | ParentNone -> false
+            | Parent p ->
+                match p.TryDeref with
+                | ValueNone -> false
+                | ValueSome e -> e.HasSignatureFile
+
+        v.IsModuleBinding && parentHasSignatureFile () && not v.HasSignatureFile
+    
     if not env.external &&
        not alreadyDone &&
        cenv.reportErrors && 
        not v.HasBeenReferenced && 
-       not v.IsCompiledAsTopLevel && 
-       not (v.DisplayName.StartsWithOrdinal("_")) && 
+       (not v.IsCompiledAsTopLevel || topLevelPublicLetBindingNotInSignatureFile ()) &&
+       not (v.DisplayName.StartsWithOrdinal("_")) &&
```

Comment:
Do we want the "ignore if starting with _" also in this case ?

Reply:
Yes, I think so. That seems consistent with how it works on the other levels.